### ISSUE TITLE
Display graphical images in terminal

### ISSUE THREAD

**Issue by JekRock** [microsoft/terminal#5746]
# Description of the new feature/enhancement
It would be nice to have the ability to display graphical images in the terminal by using the API provided by the Windows Terminal. Maybe something similar to https://sw.kovidgoyal.net/kitty/kittens/icat.html or https://www.iterm2.com/documentation-images.html

**Comment by zadjii-msft** [comment]
So we've already got #448 for display images with sixel graphics, though maybe this issue would make sense to leave open for supporting other binary image formats in the Terminal. Especially for something like gifs, which I don't think sixel is necessarily well suited for.

**Comment by DHowett-MSFT** [comment]
I'm taking triage off this. It needs a spec, and we need to make sure that we follow the prevailing standards if we're ever going to do this.

**Comment by DHowett-MSFT** [comment]
Comments from #4544.

from @zyy1998

> # Display images in terminal
> I'm trying to use **_WSL_** to substitute some functions on windows, it performed well in some text-only functions. But I'm also want to use **_ranger_** ( a linux terminal file explorer ) to replace **_totalcommander_** which I often used or even **_windows file explorer_**, **_ranger_** satisfies my content in many occasions but for image preview.
> 
> **_w3m_** is a terminal web browser that could be used to display images in terminal, and it can also be used by **_ranger_**. However, **_w3m_** didn't work on this terminal.
> 
> There are many other terminals like **_iTerm2_**、**_urxvt_** and **_alacritty_** could display images. As a terminal under windows, I think it's more necessary to support this feature. 
> 
> # Proposed technical implementation details
> I don't know how does w3m works, but I have noticed that the terminal could be changed the background pictures, so if you want to render some images, you can just change the specific area of the background of the terminal. Actually it's how **_urxvt_** handled it.
> 
> Of course, you don't have to support w3m, just do something that could display images correctly and the method could be called by other apps like **_ranger_**, that's all what I need. 
> 
> Thanks! ( And pardon for my poor English😁 )
> 
> You can find more information here: https://github.com/ranger/ranger/wiki/Image-Previews


from @akho

> w3m and überzug both rely on x11 to draw an image in a subwindow. It is highly unlikely that either of the two can be made to work in Terminal without also modifying the tools themselves.
> 
> Having some way to draw images in terminal would be very nice though. Kitty, for example, implements its own thing (icat), and integrates well with ranger. Would be lovely to have that in Terminal.
> 
> Sixel or whatever iTerm2 does are also nice, I think Sixel support is in the pipeline? Not supported by ranger yet, though.

**Comment by remkop** [comment]
This discussion between the authors of various terminal emulators may be relevant: https://gitlab.freedesktop.org/terminal-wg/specifications/issues/12

**Comment by egmontkob** [comment]
See also https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/26.

**Comment by j4james** [comment]
> See also https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/26.

https://xkcd.com/927/

**Comment by dotnetCarpenter** [comment]
I tried to use [`feh`](https://feh.finalrewind.org) in terminal but I needed X11 running. So I tried both [VcXsrv Windows X Server](https://sourceforge.net/projects/vcxsrv/) and [MobaXterm Home Edition](https://mobaxterm.mobatek.net). In both cases, I need to forward to X11 and open Windows firewall TCP port 6000 as described on [StackOverflow](https://stackoverflow.com/questions/61110603/how-to-set-up-working-x11-forwarding-on-wsl2).

But both also had a lot more than I need or want. And both will only display images in their own terminal. I want to use Microsoft terminal and I want images to look like [the mockup at wslutiliti.es (using `feh`)](https://wslutiliti.es).

I googled how to run X11 server in MS terminal and found this issue... I hope it's feasible. 🤷

**Comment by metawrap-dev** [comment]
It would be fantastic if I could just pipe out HTML5 from my console app and have it display on the terminal. 

How about that?

**Comment by zadjii-msft** [comment]
@metawrap-dev How would you propose that would actually work? How would you specify a size for the HTML content to take up within the terminal buffer? How would you deal with scenarios like ssh, where the end terminal might be on another machine entirely, or something like tmux where the end terminal isn't actually a terminal with a physical size in pixels, only in characters?

How would you suppose the terminal actually parse and render the HTML content? A WebView? Embed en entire electron instance? Is JS supported as well, or just HTML?

There's a _lot_ more questions in that suggestion than there are for just "display images", so maybe that discussion warrants its own thread.

**Comment by sintaxx** [comment]
has there been any progress on displaying images in WSL2's terminal ?

**Comment by DHowett** [comment]
no. If you'd like updates, click on the subscribe button to the right.
As long as this issue is open and there have been no comments indicating progress, you can reasonably expect that there hasn't been any progress.

**Comment by JekRock** [comment]
Here is an interesting project about rendering videos in a terminal https://github.com/ghaiklor/terminal-canvas

**Comment by dinhanhx** [comment]
Here is the closest thing I found https://github.com/nabijaczleweli/termimage

**Comment by MuizU** [comment]
Has anyone had any success using ueberzug?

**Comment by dinhanhx** [comment]
> Has anyone had any success using ueberzug?

https://github.com/microsoft/terminal/issues/4544#issuecomment-626187986

**Comment by scutxd** [comment]
> Has anyone had any success using ueberzug?

I test it, it did not work.

**Comment by zacknorman** [comment]
Just wanted to chime in as I'm excited for this... specifically being able to properly run [neofetch](https://github.com/dylanaraps/neofetch) with w3m wallpaper support... 

If it's any help, I think the escape sequence in iTerm2 would be the best case. I've always used that on MacOS and it's exceptional. [Another one worth looking into might be tycat from the terminology terminal in enlightenment](https://manpages.debian.org/unstable/terminology-data/tycat.1.en.html).

Looking forward to seeing this happen - and honestly - I'd love to help out with it.

**Comment by hhoeflin** [comment]
Here another very good program for converting images to something that can be displayed in terminal with a lot of different options.

https://github.com/hpjansson/chafa

**Comment by bragjo** [comment]
why is there no support for this yet?

**Comment by christianparpart** [comment]
> why is there no support for this yet?

Because it's complicated.

Because ConPTY needs to support that (e.g. via VT passthrough) **first** before any terminal on windows can display images genetically.

Best chances are to get it via #1173 .

**Comment by danielbisar** [comment]
A workaround is using kitty terminal on Windows 11 without running vcxsrv. Performance of the terminal is a bit slower than Windows Terminal on my machine but kittys icat is works

https://github.com/danielbisar/settings/blob/main/guides/kitty-on-windows-with-wsl2.md

**Comment by maximousblk** [comment]
any progress regarding this?

<sub>(sorry for the ping)</sub>

**Comment by koszeggy** [comment]
Apart from some well-supported future API for this it would be nice if at least the [legacy solutions](https://stackoverflow.com/a/33652557/5114784) were not broken. It has been working since Windows 95 until Windows 10 and even [ConEmu](https://conemu.github.io/en/Screenshots.html) emulates this behavior. And a sort of popular console apps such as [Far Manager](https://www.farmanager.com/download.php?l=en) could utilize it in some clever ways:
![Image in pre-Windows 11 console window](https://farplugs.sourceforge.io/img/imageview.png)

Unfortunately, this no longer works with Windows 11 Terminal. I don't know what's missing to make it work again because [`GetConsoleWindow`](https://learn.microsoft.com/en-us/windows/console/getconsolewindow) still returns the console handle as a Win32 HWND and even the [Graphics::FromHWND](https://learn.microsoft.com/en-us/windows/win32/api/gdiplusgraphics/nf-gdiplusgraphics-graphics-fromhwnd) method returns a seemingly valid instance, it's just its [VisibleClipBounds](https://learn.microsoft.com/en-us/windows/win32/api/gdiplusgraphics/nf-gdiplusgraphics-graphics-getvisibleclipbounds(rect)) is always an empty rectangle.

**Comment by odell0111** [comment]
The Python package, [image-in-terminal](https://pypi.org/project/image-in-terminal), performs admirably by replacing every two pixels of an image with the character ▀ (Upper Half Block). This makes it most suitable for displaying low-resolution images. While images with higher resolutions can also be displayed in the terminal, performance may decrease. Additionally, the terminal’s contents need to be zoomed out to view the images. This package served my small project well, and it might also be useful for others until the Windows Terminal implements support for displaying images

**Comment by lhecker** [comment]
Yes, to both questions. An initial implementation would almost certainly implement sixels first, since we already have a basic sixel parser (and related code) and it's the most basic of format. And sixels work over SSH.

**Comment by Shailosingh** [comment]
> Yes, to both questions. An initial implementation would almost certainly implement sixels first, since we already have a basic sixel parser (and related code) and it's the most basic of format. And sixels work over SSH. (Edit: And the linked formats both support SSH as well.)

Oh wow, that's great! Thanks for the update!

**Comment by iSeiryu** [comment]
> > why is there no support for this yet?
> 
> Because it's complicated.
> 
> Because ConPTY needs to support that (e.g. via VT passthrough) **first** before any other terminal on windows can display images generically.
> 
> Best chances are to get it via #1173 .

https://github.com/microsoft/terminal/issues/1173#issuecomment-1600968653
Windows Terminal Preview has a hidden setting for this now:
```json
"experimental.connection.passthroughMode": true,
```

**Comment by trackd** [comment]
having spent a fair amount of time playing around with the different Image protocols in [Sixel Module](https://github.com/trackd/Sixel)

which supports Sixel and experimental support for IIP and Kitty Graphics.

I really wish WT implemented [Kitty Graphics Protocol](https://sw.kovidgoyal.net/kitty/graphics-protocol/)

Using the image binary data like InlineImageProtocol and Kitty does is just way easier and faster.

And kitty gives you alot of control on how to render, position, animation and supports compression etc.

Sixel is really cool but feels kinda inefficient in comparison.

**Comment by ldemailly** [comment]
Having just added kitty support to my TUIs - it's really easy when you already generate PNG, basically chunk and base64 (and you don't _actually_ need to chunk) - I would love that basic support (no need to do the shm or file based/local, just the inline transmission) in Windows Terminal as well.

**Comment by j4james** [comment]
> you don't _actually_ need to chunk

Just a word of warning to anyone reading this, you definitely _do_ need to chunk. The reason chunking is mandated in the protocol is to work around an issue an issue in kitty where the screen gets flooded with garbage if you use a string sequence over a certain size. The threshold for triggering that feature was raised some time ago, so it's less likely to be a problem now, but if you create an image that's large enough, and you aren't using chunks, kitty users can still be impacted by this.
